pyscriptjs/src/runtime.ts

```diff
@@ -0,0 +1,183 @@
+import {
+    runtimeLoaded,
+    loadedEnvironments,
+    globalLoader,
+    initializers,
+    postInitializers,
+    Initializer,
+    scriptsQueue,
+    appConfig
+} from './stores'
+import type { PyScript } from './components/pyscript';
+
+// The current release doesn't export `PyodideInterface` type
+export type PyodideInterface = Awaited<ReturnType<typeof loadPyodide>>;
+export type RuntimeInterpreter = PyodideInterface | null;
+
+export type AppConfig = {
+    autoclose_loader: boolean;
+    name?: string;
+    version?: string;
+    runtimes?: Array<Runtime>;
+};
+
+let loader: PyLoader | undefined;
+globalLoader.subscribe(value => {
+    loader = value;
+});
+
+let initializers_: Initializer[];
+initializers.subscribe((value: Initializer[]) => {
+    initializers_ = value;
+    console.log('initializers set');
+});
+
+let postInitializers_: Initializer[];
+postInitializers.subscribe((value: Initializer[]) => {
+    postInitializers_ = value;
+    console.log('post initializers set');
+});
+
+let scriptsQueue_: PyScript[];
+scriptsQueue.subscribe((value: PyScript[]) => {
+    scriptsQueue_ = value;
+    console.log('scripts queue set');
+});
+
+let appConfig_: AppConfig = {
+    autoclose_loader: true,
+};
+
+appConfig.subscribe((value: AppConfig) => {
+    if (value) {
+        appConfig_ = value;
+    }
+    console.log('config set!');
+});
+
+/*
+Runtime class is a super class that all different runtimes must respect
+and adhere to.
+
+Currently, the only runtime available is Pyodide as indicated by the
+`RuntimeInterpreter` type above. This serves as a Union of types of
+different runtimes/interpreters which will be added in near future.
+
+The class has abstract methods available which each runtime is supposed
+to implement.
+
+Methods available handle loading of the interpreter, initialization,
+running code, loading and installation of packages, loading from files etc.
+
+For an example implementation, refer to the `PyodideRuntime` class
+in `pyodide.ts`
+*/
+export abstract class Runtime extends Object {
+    abstract src: string;
+    abstract name?: string;
+    abstract lang?: string;
+    abstract interpreter: RuntimeInterpreter;
+    /**
+     * global symbols table for the underlying interpreter.
+     * */
+    abstract globals: any;
+
+    /**
+     * loads the interpreter for the runtime and saves an instance of it
+     * in the `this.interpreter` property along with calling of other
+     * additional convenience functions.
+     * */
+    abstract loadInterpreter(): Promise<void>;
+
+    /**
+     * delegates the code to be run to the underlying interpreter
+     * which can call its own API behind the scenes.
+     * */
+    abstract runCode(code: string): any;
+
+    /**
+     * delegates the code to be run to the underlying interpreter
+     * (asynchronously) which can call its own API behind the scenes.
+     * */
+    abstract runCodeAsync(code: string): Promise<any>;
```

Comment:
Remove this in light of the discussion elsewhere in this PR..? (We can always add it later).

My concern is that we retain the simplest possible code base and NOT add stubs that we're likely not going to use, only to forget about them later.

Reply:
Still waiting on some consensus before we decide to remove this / merge this and non-async version.

Reply:
+1 on @ntoll suggestion

Reply:
decided on `runPythonAsync` 🗡️